openhands/controller/agent_controller.py

```diff
@@ -255,7 +261,8 @@ async def _handle_observation(self, observation: Observation):
                 self.state.metrics.merge(self.state.local_metrics)
         elif isinstance(observation, FatalErrorObservation):
             await self.report_error(
-                'There was a fatal error during agent execution: ' + str(observation)
+                'There was a fatal error during agent execution: ' + str(observation),
+                add_to_eventstream=False,
```

Comment:
why does adding it to the event stream cause the behavior?

this seems like it'd be nice to have in the EventStream (e.g. if someone does share-openhands)--is there another fix we could make?

Reply:
ohh - in the line below we actually add FatalErrorObservation directly to eventstream.

Without this fix, what happened was:
1. we add an ErrorObservtaion to eventstream - which will possibly triggers more agent actions
2. Then we add an FatalErrorObservation - which should stop the controller (but it didn't for some unknown reason)

In this PR, we simply don't do (1) - so FatalErrorObservation will properly stop the agent controller.

For ShareOpenHands -- i think as long as FatalErrorObservation is in the eventstream, we should be good, right?

Reply:
@rbren Maybe in part, because of [this](https://github.com/All-Hands-AI/OpenHands/blob/8bfd2fcf4f079ed37aca1845fa2a76c507bf6db7/openhands/agenthub/codeact_agent/codeact_agent.py#L172). We currently treat all error obs the same, and send them to the agent.

Alternative would be [like this](https://github.com/All-Hands-AI/OpenHands/pull/3808/files#diff-1e082f415afb40de1d5c2a3fa327326225e1da28a142ad89c9524109f35fb36eR72). On current `main` I think adding the Fatal obs [here](https://github.com/All-Hands-AI/OpenHands/blob/8bfd2fcf4f079ed37aca1845fa2a76c507bf6db7/openhands/memory/history.py#L35) should do the same.

Reply:
Wait, we also should reraise this Fatal, maybe [here](https://github.com/All-Hands-AI/OpenHands/blob/8bfd2fcf4f079ed37aca1845fa2a76c507bf6db7/openhands/controller/agent_controller.py#L261).

Although... ideally we wouldn't use send_error for it. It has too many purposes...

Reply:
Ahh the real problem here I think is that we're adding it to the event stream in a loop! This is an event stream handler, and it's adding the same event its handling back into the event stream

Very tricky. https://github.com/All-Hands-AI/OpenHands/pull/4575 for a better failure here

Reply:
Agreed `report_error` is doing too much work here. Not sure why we're tracking `last_error` to begin with--it doesn't seem to be used anywhere

Reply:
It's used in evals, it will be part of the final report.

Reply:
@xingyaoww I see several other places where we _first_ set agent state, _then_ `report_error`, along with a note that this is important.

Can you try making that change? i.e. just move line 267 up, and revert everything else?

I'm going to work on making `report_error` better on the side

Reply:
Actually I think while you're right in essence, Robert, I'm not sure we are indeed adding the same obs.

> which should stop the controller (but it didn't for some unknown reason)

TBH I am confused by the error handling in the controller lately and I'm again not so sure about this one. Allow me to go back to basics. These are observations, not Exceptions. The way it would stop the controller is if we get [here in the start loop](https://github.com/All-Hands-AI/OpenHands/blob/8bfd2fcf4f079ed37aca1845fa2a76c507bf6db7/openhands/controller/agent_controller.py#L170) with an exception, is that correct? Is there another way to stop than an uncaught exception? Can't we just raise on a FatalObservation?

Reply:
@rbren 
> Can you try making that change? i.e. just move line 267 up, and revert everything else?

Just pushed a most recent commit.

Reply:
and I think currently the controller doesn't stop hard even when the agent got stuck in a loop 

<img width="1684" alt="image" src="https://github.com/user-attachments/assets/8f437294-b910-4216-88d0-d49b84b8db87">